pkg/grpc/jwt_authenticator_test.go

```diff
@@ -0,0 +1,234 @@
+		md := metadata.Pairs("authorization", "Bearer "+token)
+		require.NoError(
+			t,
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+
+	t.Run("RejectsInvalidJWS", func(t *testing.T) {
+		// Should reject an invalid JWS.
+		invalidPartsSignedToken := `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJzdWJqZWN0IiwiaXNzIjoiaXNzdWVyIiwic2NvcGVzIjpbInMxIiwiczIiXX0`
+		md := metadata.Pairs("authorization", "Bearer "+invalidPartsSignedToken)
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Failed to parse signed bearer token: square/go-jose: compact JWS format must have three parts"),
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+
+	t.Run("RejectsExpiredJWS", func(t *testing.T) {
+		// Should reject an expired JWS.
+		clock.EXPECT().Now().Return(time.Unix(1600000000, 0))
+		token := "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1OTk5OTY0MDAsImlzcyI6ImJ1aWxkYmFybiIsInN1YiI6InN1YmplY3QifQ.tLxc6KRzPI7qL5wCl7eL-iLguinJzKnO11ejBqxsrE4"
+		md := metadata.Pairs("authorization", "Bearer "+token)
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Authorization required: square/go-jose/jwt: validation failed, token is expired (exp)"),
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+
+	t.Run("ParsesAndValidateValidJWSWithPublicKey", func(t *testing.T) {
+		input := []byte(rsaPublicKey)
+		block, _ := pem.Decode(input)
+		if block != nil {
+			input = block.Bytes
+		}
+
+		rsaPubKey, err := x509.ParsePKIXPublicKey(input)
+		require.NoError(t, err)
+
+		jwtKey := bb_grpc.JWTKeyConfig{
+			Key: rsaPubKey,
+		}
+
+		authenticator := bb_grpc.NewJWTAuthenticator(jwtKey, clock)
+
+		clock.EXPECT().Now().Return(time.Unix(1600000000, 0))
+		token := "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJidWlsZGJhcm4iLCJzdWIiOiJzdWJqZWN0In0.JZgfFQW47GWEeN6Esn40gnM4XA4QiWXq8ejAWkPIup5b7K8PGeVjnNfGGZhQsI6xmkKDFO4qbRbnJByhtDCYcPNfNAVz3iA_zXnBXKucY5jFpkhM7mvkr55NVDDkVSkyW5tjz3vTXTUMko5UgPJ5RLlZFrTGRYNRu5h8dxGc5MD5O46bKmQbn9JjV8fT6ngSXjJjC06KMjDLDutYyyTLj2hIeFbeQgIfKClYqmrH47XMSeyOHH1-bCmdWmZVwQhwsjYaGBHy2-SxkwocKOBB9Wo_8lpl301PogX1sqIlIE1bt_aylHZbT27l8gTov51r8qSJzhZ30CSgAWFysH8rHg"
+
+		md := metadata.Pairs("authorization", "Bearer "+token)
+		require.NoError(
+			t,
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+}
+
+func decodeRsaPrivateKey(privateKeyData string) *rsa.PrivateKey {
+	block, _ := pem.Decode([]byte(privateKeyData))
+	if block == nil {
+		panic("failed to decode PEM data")
+	}
+	key, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		panic("failed to parse RSA key: " + err.Error())
+	}
+	return key
+}
+
+// Generates the JWT test fixtures used in other tests in this file. Uncomment the
+// `t.Skip` call in the test, run the test, and then copy the test output to the
+// appropriate tests above.
+func TestGenerateJWTAuthenticatorFixtures(t *testing.T) {
```

Comment:
This "test" generates the JWT fixtures for the `TestJWTAuthenticator` test. The test is marked skipped so it does not execute normally and is not commented out so that it will still be compiled to avoid bit rot.